core/types/calls.cc

```diff
@@ -2919,6 +2937,81 @@ class Magic_mergeHashValues : public IntrinsicMethod {
     }
 } Magic_mergeHashValues;
 
+namespace {
+void convertDefinedToType(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) {
+    // args[0] is the receiver
+    // args[1] is the symbol name of the variable (class or instance)
+    if (args.args.size() != 2) {
+        return;
+    }
+
+    auto &receiver = args.args[0];
+    if (receiver->type.isUntyped()) {
+        res.returnType = receiver->type;
+        return;
+    }
+    if (!receiver->type.isFullyDefined()) {
+        return;
+    }
+
+    auto &classVar = args.args[1];
+    auto classLit = cast_type_nonnull<LiteralType>(classVar->type);
+    auto classVarName = classLit.asName(gs);
+
+    SymbolRef klass;
+    if (isa_type<AppliedType>(receiver->type)) {
+        klass = cast_type_nonnull<AppliedType>(receiver->type).klass;
+        klass = klass.data(gs)->attachedClass(gs);
+        ENFORCE(klass.exists());
+    } else if (isa_type<ClassType>(receiver->type)) {
+        klass = cast_type_nonnull<ClassType>(receiver->type).symbol;
+
+    } else {
+        return;
+    }
+
+    auto selfData = klass.data(gs);
+    auto var = selfData->findMemberTransitive(gs, classVarName);
+    if (!var.exists()) {
+        return;
+    }
+
+    auto &symType = var.data(gs)->resultType;
+    if (symType == nullptr) {
+        return;
+    }
+
+    res.returnType = symType;
+}
+} // namespace
+
+class Magic_definedInstanceVar : public IntrinsicMethod {
+    // We want `<Magic>.<defined-instance-var>(self, :varname)` to appear to inference
+    // as `@varname` so things like `if <Magic>.<defined-instance-var>(...)` will potentially
+    // warn about unreachable code.
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        convertDefinedToType(gs, args, res);
+    }
+} Magic_definedInstanceVar;
```

Comment:
Does this PR silence the error that Sorbet currently reports for this snippet?

```ruby
# typed: strict

class A
  @@x ||= @@x
end
```

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0A%0Aclass%20A%0A%20%20%40%40x%20%7C%7C%3D%20%40%40x%0Aend)

```
editor.rb:4: Use of undeclared variable @@x https://srb.help/6002
     4 |  @@x ||= @@x
          ^^^
Errors: 1
```

Reply:
if not, do we have a test case like this already?